### PR TITLE
Allow precision for %p printf specifiers for CHERI.

### PR DESCRIPTION
**Pull request by bsdjhb**

### CODE REVIEW

**Comment by arichardson** [comment]
Seems fine to me since it looks like we do the same for "#". However, I do wonder if we should restrict it to purecap and/or FreeBSD triple since other libcs might print nonsense values?

A test would also be nice (especially if we restrict where this is allowed).

**Comment by bsdjhb** [comment]
I wonder if we should block %p for scanf() if it isn't already blocked?  That's for a separate day though.

**Comment by bsdjhb** [comment]
@arichardson are you happier with this version that restricts the additions to CHERI targets?

File: clang/lib/AST/PrintfFormatString.cpp [hunk 1]
```diff
@@ -1111,6 +1116,9 @@ bool PrintfSpecifier::hasValidPrecision() const {
   case ConversionSpecifier::FreeBSDyArg:
   case ConversionSpecifier::PArg:
     return true;
+  case ConversionSpecifier::CHERIpArg:
```
Comment (jrtc27) [inline clang/lib/AST/PrintfFormatString.cpp:1119]: Maybe
```suggestion

  case ConversionSpecifier::CHERIpArg:
```
given the blank line below?

File: clang/lib/AST/FormatString.cpp [hunk 2]
```diff
@@ -810,6 +811,7 @@ bool FormatSpecifier::hasValidLengthModifier(const TargetInfo &Target,
         case ConversionSpecifier::sArg:
         case ConversionSpecifier::ScanListArg:
         case ConversionSpecifier::ZArg:
+        case ConversionSpecifier::CHERIpArg:
```
Comment (arichardson) [inline clang/lib/AST/FormatString.cpp:814]: This should probably be restricted to hybrid/only allow for `l`?

File: clang/test/Sema/format-strings-cheri.c [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  printf("%#.4p", p);
+}
+
+#ifndef __CHERI_PURE_CAPABILITY__
+void test_hybrid(void * __capability p) {
+  printf("%#lp", &p);
+  printf("%.4lp", &p); // expected-warning{{precision used with 'p' conversion specifier, resulting in undefined behavior}}
+  printf("%#.4lp", &p);
+}
+#endif
```
Comment (arichardson) [inline clang/test/Sema/format-strings-cheri.c:18]:
```suggestion
```
Should probably be a separate commit that adds the l flag.

File: clang/lib/AST/FormatString.cpp (hunk 2 above)
Comment (jrtc27) [reply clang/lib/AST/FormatString.cpp:814]: This is within `LengthModifier::AsLong`?

File: clang/test/Sema/format-strings-cheri.c (hunk 3 above)
Comment (arichardson) [reply clang/test/Sema/format-strings-cheri.c:18]: It would also be nice to check that other length modifiers warn.

File: clang/lib/AST/FormatString.cpp (hunk 2 above)
Comment (bsdjhb) [reply clang/lib/AST/FormatString.cpp:814]: Hmm, I had assumed it would be hard to restrict this to hybrid, but since TargetInfo is present perhaps I can make that work.  Would need to figure out how to make the tests DTRT.

Comment (jrtc27) [reply clang/lib/AST/FormatString.cpp:814]: It depends whether that's what we want. Having to use:

- `%p` for non-CHERI
- `%#lp` for hybrid
- `%#p` for purecap

would be rather annoying; permitting a redundant `l` in `%#lp` for purecap would help (and there's not much more you can do to help that given no other compiler/libc will let you write `%#lp`... well, some might ignore the characters and let you get away with it, but they don't have to).

**Comment by bsdjhb** [comment]
Ok, this is now 3 commits and a bit cleaner I think.  The first one introduces CHERIpArg and makes the existing `#` special case only applied to CHERI.  The second permits using precision with `#` for CHERIpArg.  The third permits using `l` with CHERI hybrid (but not purecap, the function has the Target around, so this was easy).  The tests are also a bit expanded to test that all the other length modifiers with `%p` fail, and that `%lp` fails for purecap.

**Review by arichardson — COMMENTED** [review]
Generally LGTM, just a suggestion to avoid ifdefs. 
Just to confirm `#lp` expects `&cap` as the argument? I guess we could make that also work for purecap. Do we want to?

**Review by arichardson — APPROVED** [review]

File: clang/lib/AST/FormatString.cpp (hunk 2 above)
Comment (bsdjhb) [reply clang/lib/AST/FormatString.cpp]: Hmm, I'm fine permitting 'lp' for purecap.  I guess it's akin to how we permit `__capability` in purecap as well.

**Comment by jrtc27** [comment]
> Generally LGTM, just a suggestion to avoid ifdefs.
> Just to confirm `#lp` expects `&cap` as the argument? I guess we could make that also work for purecap. Do we want to?

I think for simplicity, yes, we should make that work in the same way as hybrid for purecap, and then once we fix hybrid varargs to pass capabilities indirectly then it can be that `#lp` works everywhere with `cap`?

**Comment by bsdjhb** [comment]
Just fixed 'lp' to also work for purecap.  (I'll fix the CheriBSD changes to permit that if needed, but I think it already "works")
